dev_scripts/qa.py

```diff
@@ -0,0 +1,448 @@
+                    choice = self.prompt(ref=ref, prompt=prompt)
+                    try_auto = "auto" == choice
+
+                # Run the task automatically, if the user asked to, else assume that the
+                # user has run it manually.
+                if try_auto:
+                    logger.info("Trying automatically...")
+                    out = func(self, *args, **kwargs)
+                    logger.info("Completed successfully")
+                    return out
+                else:
+                    logger.debug("Assuming step %s has completed manually...", func)
+
+            return inner
+
+        return decorator
+
+    def run(self, *args):
+        """Run a command with extra debug logs."""
+        # Construct the command line representation of the provided arguments, by taking
+        # the arguments and joining them with a space character. If an argument already
+        # has a space character in it, quote it before adding it to the CLI
+        # representation.
+        cmd_str = ""
+        for arg in args:
+            if " " in arg:
+                cmd_str += " '" + arg + "'"
+            else:
+                cmd_str += " " + arg
+
+        logger.debug("Running %s", cmd_str)
+        try:
+            subprocess.run(args, check=True)
+            logger.debug("Successfully ran %s", cmd_str)
+        except subprocess.SubprocessError as e:
+            logger.debug("Failed to run %s", cmd_str)
+            raise
+
+    def try_run(self, *args):
+        """Try to run a command and return True/False, depending on the result."""
+        try:
+            self.run(*args)
+            return True
+        except subprocess.SubprocessError as e:
+            return False
+
+    def describe(self, *msgs, ref=None):
+        """Print a task's description.
+
+        Log one message per line, and optionally allow a URL as a reference.
+        """
+        for msg in msgs:
+            logger.info(msg)
+        if ref:
+            logger.info("For reference, see: %s" % ref)
+
+    def prompt(self, *msgs, ref=None, prompt=None):
+        """Print the task's description, and prompt the user for an action."""
+        self.describe(*msgs, ref=ref)
+        prompt = prompt or "Press Enter once you completed this step to continue: "
+        if self.skip_manual:
+            logger.info("Skipping manual tasks, as instructed")
+            return None
+        else:
+            return input(prompt)
+
+    def qa_scenarios(self, skip=None):
+        """Common prompt for the QA scenarios.
+
+        This method suggests to the user a way to check the QA scenarios, in a
+        Dangerzone environment. Then, it iterates through them, prints their
+        description, and asks the user if they pass.
+        """
+        skip = skip or []
+        REF_QA = "https://github.com/freedomofpress/dangerzone/blob/49b7736cb49d8b03d44c66a3903ba4b9ba1f9171/RELEASE.md#scenarios"
```

Comment:
I agree that the scenarios shouldn't be out of sync as well, but I'd prefer if users could see them, instead of finding them from a dev script. They are useful as they can see at a glance if a change they want to make must be tested also manually, especially if it's a GUI one.

In any case, we can cover these deviations with the solution you provided above.

Reply:
Yes, let's do that.